### Нейросеть для распознавания рукописных цифр. Dataset Mnist.

In [1]:
import numpy as np
import tensorflow as tf

In [2]:
tf.__version__

'2.1.0'

In [3]:
print("Num GPUs Available: ", len(tf.config.experimental.list_physical_devices('GPU')))

Num GPUs Available:  1


In [4]:
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical

Using TensorFlow backend.


In [5]:
# loading dataset
(train_images, train_labels), (test_images, test_labels) = tf.keras.datasets.mnist.load_data()

In [6]:
# Normalize the images.
train_images = (train_images / 255) - 0.5
test_images = (test_images / 255) - 0.5

In [7]:
# Flatten the images.
train_images = train_images.reshape((-1, 784))
test_images = test_images.reshape((-1, 784))

In [8]:
# Build the model.
model = Sequential([
  Dense(64, activation='relu', input_shape=(784,)),
  Dense(64, activation='relu'),
  Dense(10, activation='softmax'),
])

In [9]:
# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

In [10]:
# Train the model.
model.fit(
    train_images,
    to_categorical(train_labels),
    epochs=5,
    batch_size=32,
    )

Epoch 1/5
60000/60000 [==============================] - 2s 40us/step - loss: 0.3734 - accuracy: 0.8890
Epoch 2/5
60000/60000 [==============================] - 2s 37us/step - loss: 0.2396 - accuracy: 0.9290
Epoch 3/5
60000/60000 [==============================] - 2s 37us/step - loss: 0.1946 - accuracy: 0.9412
Epoch 4/5
60000/60000 [==============================] - 2s 37us/step - loss: 0.1650 - accuracy: 0.9508
Epoch 5/5
60000/60000 [==============================] - 2s 37us/step - loss: 0.1434 - accuracy: 0.9565


In [11]:
# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

10000/10000 [==============================] - 0s 31us/step


[0.14451309633888304, 0.9563999772071838]

In [12]:
# Save the model to disk.
model.save_weights('model.h5')

# Load the model from disk later using:
# model.load_weights('model.h5')

In [13]:
# Predict on the first 5 test images.
predictions = model.predict(test_images[:7])

In [14]:
# Print our model's predictions.
print(np.argmax(predictions, axis=1)) # [7, 2, 1, 0, 4]

[7 2 1 0 4 1 4]


In [15]:
# Check our predictions against the ground truths.
print(test_labels[:7]) # [7, 2, 1, 0, 4]

[7 2 1 0 4 1 4]


__Увеличим кол-во внутренних слоёв до 3-х   
Кол-во нейронов внутренних слоях: 256, 192, 128   
Кол-во эпох увеличим до 12-ти   
batch_size=64__

In [16]:
model = Sequential([
  Dense(256, activation='relu', input_shape=(784,)),
  Dense(192, activation='relu'),
  Dense(128, activation='relu'),    
  Dense(10, activation='softmax'),
])

In [17]:
model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy'],
)

In [18]:
model.fit(
    train_images,
    to_categorical(train_labels),
    epochs=12,
    batch_size=64,
    )

Epoch 1/12
60000/60000 [==============================] - 1s 22us/step - loss: 0.2911 - accuracy: 0.9093
Epoch 2/12
60000/60000 [==============================] - 1s 22us/step - loss: 0.1383 - accuracy: 0.9572
Epoch 3/12
60000/60000 [==============================] - 1s 22us/step - loss: 0.1047 - accuracy: 0.9680
Epoch 4/12
60000/60000 [==============================] - 1s 21us/step - loss: 0.0852 - accuracy: 0.9736
Epoch 5/12
60000/60000 [==============================] - 1s 21us/step - loss: 0.0705 - accuracy: 0.9779
Epoch 6/12
60000/60000 [==============================] - 1s 21us/step - loss: 0.0596 - accuracy: 0.9813
Epoch 7/12
60000/60000 [==============================] - 1s 21us/step - loss: 0.0525 - accuracy: 0.9837
Epoch 8/12
60000/60000 [==============================] - 1s 21us/step - loss: 0.0454 - accuracy: 0.9857
Epoch 9/12
60000/60000 [==============================] - 1s 21us/step - loss: 0.0399 - accuracy: 0.9870
Epoch 10/12
60000/60000 [==============================

In [19]:
model.evaluate(test_images, to_categorical(test_labels))

10000/10000 [==============================] - 0s 32us/step


[0.07990907403648016, 0.9775999784469604]

Точность нейронной сети на тестовых данных увеличилась с 0.9564 до 0.9776

#### Изменение оптимизатора

In [20]:
model = Sequential([
  Dense(256, activation='relu', input_shape=(784,)),
  Dense(192, activation='relu'),
  Dense(128, activation='relu'),    
  Dense(10, activation='softmax'),
])

In [21]:
from keras import optimizers
sgd = optimizers.SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(
    optimizer=sgd,
    loss='categorical_crossentropy',
    metrics=['accuracy'],
)

In [22]:
model.fit(
    train_images,
    to_categorical(train_labels),
    epochs=12,
    batch_size=64,
    )

Epoch 1/12
60000/60000 [==============================] - 1s 21us/step - loss: 0.3416 - accuracy: 0.8964
Epoch 2/12
60000/60000 [==============================] - 1s 20us/step - loss: 0.1489 - accuracy: 0.9551
Epoch 3/12
60000/60000 [==============================] - 1s 20us/step - loss: 0.1079 - accuracy: 0.9663
Epoch 4/12
60000/60000 [==============================] - 1s 20us/step - loss: 0.0849 - accuracy: 0.9731
Epoch 5/12
60000/60000 [==============================] - 1s 21us/step - loss: 0.0710 - accuracy: 0.9772
Epoch 6/12
60000/60000 [==============================] - 1s 21us/step - loss: 0.0580 - accuracy: 0.9815
Epoch 7/12
60000/60000 [==============================] - 1s 21us/step - loss: 0.0513 - accuracy: 0.9832
Epoch 8/12
60000/60000 [==============================] - 1s 21us/step - loss: 0.0421 - accuracy: 0.9863
Epoch 9/12
60000/60000 [==============================] - 1s 20us/step - loss: 0.0376 - accuracy: 0.9875
Epoch 10/12
60000/60000 [==============================

In [23]:
model.evaluate(test_images, to_categorical(test_labels))

10000/10000 [==============================] - 0s 32us/step


[0.08278283290118561, 0.9758999943733215]

#### Сохранение/загрузка только архитектуры модели

In [24]:
# сохранить как JSON
json_string = model.to_json()

In [25]:
json_string

'{"class_name": "Sequential", "config": {"name": "sequential_3", "layers": [{"class_name": "Dense", "config": {"name": "dense_8", "trainable": true, "batch_input_shape": [null, 784], "dtype": "float32", "units": 256, "activation": "relu", "use_bias": true, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"scale": 1.0, "mode": "fan_avg", "distribution": "uniform", "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}}, {"class_name": "Dense", "config": {"name": "dense_9", "trainable": true, "dtype": "float32", "units": 192, "activation": "relu", "use_bias": true, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"scale": 1.0, "mode": "fan_avg", "distribution": "uniform", "seed": null}}, "bias_initializer": {"class_name": "Zeros", "config": {}}, "kernel_regularizer": null, "bias_regularizer": null, 

In [26]:
# сохранить как YAML
yaml_string = model.to_yaml()
print(yaml_string)

backend: tensorflow
class_name: Sequential
config:
  layers:
  - class_name: Dense
    config:
      activation: relu
      activity_regularizer: null
      batch_input_shape: !!python/tuple
      - null
      - 784
      bias_constraint: null
      bias_initializer:
        class_name: Zeros
        config: {}
      bias_regularizer: null
      dtype: float32
      kernel_constraint: null
      kernel_initializer:
        class_name: VarianceScaling
        config:
          distribution: uniform
          mode: fan_avg
          scale: 1.0
          seed: null
      kernel_regularizer: null
      name: dense_8
      trainable: true
      units: 256
      use_bias: true
  - class_name: Dense
    config:
      activation: relu
      activity_regularizer: null
      bias_constraint: null
      bias_initializer:
        class_name: Zeros
        config: {}
      bias_regularizer: null
      dtype: float32
      kernel_constraint: null
      kernel_initializer:
        class_name: Varianc

In [27]:
# Генерируемые JSON / YAML файлы читаются человеком и при необходимости могут быть отредактированы вручную.

# Из этих данных можно построить свежую модель:
# реконструкция модели из JSON:
from keras.models import model_from_json
model_new = model_from_json(json_string)

In [28]:
# Можно прервать обучение, когда потеря валидации больше не уменьшается - использовать обратный вызов EarlyStopping:

# from keras.callbacks import EarlyStopping
# early_stopping = EarlyStopping(monitor='val_loss', patience=2)
# model.fit(x, y, validation_split=0.2, callbacks=[early_stopping])

In [29]:
# Последний добавленный слой в последовательной модели можно удалить с помощью вызова .pop():

model = Sequential()
model.add(Dense(32, activation='relu', input_dim=784))
model.add(Dense(32, activation='relu'))
print(len(model.layers))  # «2»

2


In [30]:
model.pop()
print(len(model.layers))  # «1»

1
